0010-keccak-syscall.md

```diff
@@ -0,0 +1,402 @@
+and `a3` must be set to `0`.
+
+### `sysBrk`
+
+<!-- DONE -->
+
+The register `v0` must be set to `0x400000` and the register `a3` to `0`.
+
+### `sysClone`
+
+<!-- DONE -->
+
+The register `v0` must be set to `1` and the register `a3` to `0`.
+
+### `sysExitGroup`
+
+<!-- DONE -->
+
+The program must halt in this case and `set_halted` must be called with the value `a0`.
+
+### `sysRead`
+
+This syscall is used to get data from the outside world and write in the VM memory.
+
+The register `a0` contains the file descriptor of the input channel, the
+register `a1` the address and the `a2` contains the number of bytes that must be
+read.
+
+At the end of the call, the register `v0` must contain the number of bytes that
+have been read, and `a3` must contain the error code.
+
+Different input channels must be supported.
+
+#### Standard input
+
+<!-- DONE -->
+
+In the case of the standard input, i.e. when `a0` equals `0`, nothing in
+particular happens. Therefore, the registers `v0` and `a3` are set to `0`, which
+corresponds to no data has been read and no error is raised.
+
+#### Pre-image oracle
+
+When the buffer to be read is the pre-image oracle, i.e. when `a0` equals `5`, a
+chunk of the pre-image is read, and copy into the memory.
+The pre-image that must be read is determined by the `preimageKey`, and the actual
+chunk is given by the preimage offset variable.
+
+The number of bytes that must be read is in the register `a2` and the address
+where the chunk as to be written is in the register `a1`.
+
+At the end of the call, the register `v0` must contain the number of bytes that
+have been read. Also, the preimage offset has to be incremented by the number of
+bytes that has been read.
+
+The preimage data length is located in the 8 first addresses, encoded on 64
+bits in big-endian.
+
+A difference with Cannon is we verify the preimage data is the actual data the
+user has requested previously. To achieve this, we will use the optimised
+keccak gate described in [Keccak RFC](./keccak.md).
+
+The keccak sponge state is 136 bytes, and is located in the columns `[100,
+168[`. While reading the preimage data, we can fill at the same time the sponge
+state. When 136 bytes have been read, a keccak permutation is executed.
+If we have finished reading the data, we run the permutation. In the case the total
+length is a multiple of 136, an additional permutation is executed
+with the expected padding.
+
+Reading additional bytes after the EOF must not have any effect and must be
+allowed. This can be achieved by activating the "preimage read gate" iff the
+different between the number of bytes that have to be read and the preimage
+offset is not zero. In the case we arrived at the end of the file, the preimage
+offset must not be increased and the number of bytes read must be set `0`.
+
+After we read the data and run the permutation a last time as described above,
+we must verify that the preimage key corresponds to the output of sponge.
```

Comment:
Mmmm trying to connect the dots here...

So this means, when you were reading preimage keys before, you wanted to store them in a table and I didn't understand why. Is this supposed to be the reason? After running keccak, you want to check that the log in the table for this file corresponds (without the first byte) to the output of the keccak gadget? Is this the motivation behind the "communication channel"?

If so, is it a runtime table where you write terms and right after when executing keccak, you remove from it? With the +/- technique?